lib/ansible/modules/windows/win_xml.ps1

```diff
@@ -0,0 +1,191 @@
+    Fail-Json $result "Specified path $dest does not exist or is not a file."
+}
+
+[xml]$xmlorig = Get-Content -Path $dest
+$xmlnew = $xmlorig.Clone()
+if ($type -eq "element") {
+    $xmlchild = [xml]$fragment
+
+    $child = $xmlnew.CreateElement($xmlchild.DocumentElement.Name, $xmlnew.$root.NamespaceURI)
+    foreach ($attr in $xmlchild.DocumentElement.Attributes) {
+        $child.SetAttribute($attr);
+    }
+
+    foreach ($node in $xmlchild.DocumentElement.ChildNodes) {
+        $newnode = $xmlnew.CreateElement($node.Name, $xmlnew.$root.NamespaceURI)
+        foreach ($attr in $node.Attributes) {
+            $newnode.SetAttribute($attr)
+        }
+        $newnode.InnerText = $node.InnerText
+        $child.AppendChild($newnode) | Out-Null
+    }
+
+    $elements = FindElementsByName $xmlnew.$root $xmlchild.DocumentElement.Name
+
+    [bool]$add = $TRUE
+    if ($elements.Count) {
+        foreach ($element in $elements) {
+            try {
+                Compare-XmlDocs $xmlchild.DocumentElement $element
+                $add = $FALSE
+                break
+            } catch {
+                $_.InvocationInfo | Out-File c:\win_xml.log
+            }
+        }
+        if ($add) {
+            [void]$xmlnew.$root.AppendChild($child)
+        }
+    } else {
+        [void]$xmlnew.$root.AppendChild($child)
+    }
+
+    if ($add) {
+        $result.changed = $TRUE
+        if (-Not $params._ansible_check_mode) {
+            if ($backup -eq "yes") {
+                $result.backup = BackupFile($dest)
+            }
+            $result.msg = "xml added"
+            $xmlnew.Save($dest)
+        } else {
+            $result.msg = "added check mode"
+        }
+    } else {
+        $result.msg = "already present"
+    }
+} elseif ($type -eq "text") {
+    $node = $xmlnew.SelectSingleNode($root)
+    [bool]$add = ($node.InnerText -ne $fragment)
+    if ($add) {
+        $result.changed = $TRUE
+        if (-Not $params._ansible_check_mode) {
+            if ($backup -eq "yes") {
+                $result.backup = BackupFile($dest)
+            }
+            $node.InnerText = $fragment
+            $xmlnew.Save("$dest")
+            $result.msg = "text changed"
+        } else {
+            $result.msg = "text changed check mode"
+        }
+    } else {
+        $result.msg = "not changed"
+    }
+} elseif ($type -eq "attribute") {
+    $node = $xmlnew.SelectSingleNode($root)
+    [bool]$add = !$node.HasAttribute($attribute) -Or ($node.$attribute -ne $fragment)
+    if ($add) {
+        $result.changed = $TRUE
+        if (-Not $params._ansible_check_mode) {
+            if ($backup -eq "yes") {
+                $result.backup = BackupFile($dest)
+            }
+	    if (!$node.HasAttribute($attribute)) {
+	        $node.SetAttributeNode($attribute, $xmlnew.$root.NamespaceURI)
+            }
+            $node.SetAttribute($attribute, $fragment)
+            $xmlnew.Save("$dest")
```

Comment:
Only the `$xmlnew.Save($dest)` needs to be skipped in check-mode. We really want to do as much as possible in check-mode except the change itself. So that if any condition, or problem is found in check-mode as well (usually before performing it in production). So the more it can tests, the better ! (That's also why we prefer `-WhatIf` support in cmdlets and functions, so we leave it up to the cmdlet how far it can take this)